### PR TITLE
[core] enhance `ray.remote` type hints for actor class

### PR DESCRIPTION
**Pull request by zen-xu**
## Why are these changes needed?

It is very painful to manipulate actors because the Actor class decorated with `ray.remote` completely loses its class information, making it impossible for the IDE to provide autocompletion.

Furthermore, if we want to find where the actor's methods are called, we can only rely on text matching.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by zen-xu** [comment]
## This PR enhances type hints for actors.

2031689585465_ pic

2041689585482_ pic

2051689585510_ pic

### A Better way to call actor method

we can get all actor methods from `actor.methods`

2061689585568_ pic

pass the method to `actor.remote` as its first param, then pass the method args and will get `ObjectRef[ReturnType]`

2071689585595_ pic

if we pass wrong method arg, mypy will give error message

2081689585926_ pic

And I strongly discourage using `actor_class` as a decorator. Because the class returned by actor_class adds methods like remote that are not present in the original class, it still causes mypy errors.

2091689586382_ pic

we can declare a new remote actor class.
WeChat11b3d129c42077c4805e85b190e85807

**Comment by krfricke** [comment]
Is there a reason why we can't extend the current `@ray.remote` API to have better typehints? I don't think we should introduce a new first-class API or rename the existing one.

**Comment by zen-xu** [comment]
And if you only want to match type，then you will lose the class init signature

**Comment by zen-xu** [comment]
I got a new idea

we can add a flag boolean param `actor_class` to `ray.remote` to distinguish function or class

```python
@ray.remote(actor_class=True)
class C:
    def __init__(self, a: int):
        ...
```

**Comment by zen-xu** [comment]
now we can have better type hints by using `ray.remote(actor_class=True)`
WeChat142842f9413573fa520f372e1fc403fe

**Comment by zen-xu** [comment]
The difference between `actor_class=True` and `actor_class=False`

image

image

**Comment by zen-xu** [comment]
@richardliaw Could you spare some time to review my code?

File: python/ray/_private/worker.py [hunk 1]
```diff
@@ -2829,7 +2837,14 @@ def _make_remote(function_or_class, options):
     if not function_or_class.__module__:
         function_or_class.__module__ = "global"
 
+    actor_class: bool = options.pop("actor_class", False)
     if inspect.isfunction(function_or_class) or is_cython(function_or_class):
+        if actor_class:
+            raise TypeError(
+                "The @ray.remote decorator must be applied "
+                "to a class when actor_class=True"
```
Comment (jjyao) [inline python/ray/_private/worker.py:2845]: This is backward incompatible.

**Review by jjyao — COMMENTED** [review]
This PR changes many Ray public APIs.

Is there a way to add type hints or auto-complete without changing the Ray APIs?

**Comment by zen-xu** [comment]
# What changes

- `ray.remote` add a new param `actor_class` to distinguish a function or a class

- `ActorHandle` add a property `methods` return itself to disguise as the actor class

- `ActorHandle` add a method `remote` for a new style remote call

File: python/ray/_private/worker.py (hunk 1 above)
Comment (zen-xu) [reply python/ray/_private/worker.py:2845]: Can you accept the original proposal and add a new `actor_class` decorator？

Comment (zen-xu) [reply python/ray/_private/worker.py:2845]: Regardless, these modifications theoretically do not affect the previous code. If the user accepts this new style remote call, it implies that they accept updating all ray versions to the latest one.

**Comment by rkooo567** [comment]
We can't accept the contribution if it requires huge public API changes.

**Comment by rkooo567** [comment]
I am 100% on improving type hints and autocompletion, but I don't think it is possible to pass the API review and move forward with this PR when we have to introduce actor_class and `methods` property. Is there any way to achieve the same thing without introducing them?

**Comment by zen-xu** [comment]
> I am 100% on improving type hints and autocompletion, but I don't think it is possible to pass the API review and move forward with this PR when we have to introduce actor_class and property. Is there any way to achieve the same thing without introducing them`methods`?

Unfortunately, when a class is decorated by `ray.remote`, all information is lost, and it becomes impossible to access legitimate methods through an IDE.

image

Otherwise, we could have accessed the legitimate methods through the class rather than the property `methods`.

image

**Comment by zen-xu** [comment]
There is a non-intrusive way, but it looks very cumbersome.

image

**Comment by rkooo567** [comment]
I think we should discuss 1 by 1 here. There are 2 APIs you propose (ray.actor_class, and methods)

What's the fundamental limitation we cannot just make ray.remote to support this? If @ray.actor_calss works, there's no reason why ray.remote cannot support the same thing? Isn't this just the problem from the ray.remote implementation?

**Comment by zen-xu** [comment]
> I think we should discuss 1 by 1 here. There are 2 APIs you propose (ray.actor_class, and methods)
> 
> What's the fundamental limitation we cannot just make ray.remote to support this? If @ray.actor_calss works, there's no reason why ray.remote cannot support the same thing? Isn't this just the problem from the ray.remote implementation?

Because we cannot determine whether it is a function or a class based on the signature.

image

**Comment by zen-xu** [comment]
Unfortunately, there might not be a possibility of merging.

**Pull request closed without merging**